File: scripts/algorithms/C/Concatenation of Consecutive Binary Numbers/Concatenation of Consecutive Binary Numbers.js
var concatenatedBinary = function(n) {
    let num = 0;
    
    for(let i = 1; i <= n; i++) {
        //OR num *= (1 << i.toString(2).length)
        num *= (2 ** i.toString(2).length) 
        num += i;
        num %= (10 ** 9 + 7)
    }
    return num;
};

File: scripts/algorithms/C/Concatenation of Consecutive Binary Numbers/Concatenation of Consecutive Binary Numbers.py
class Solution:
    def concatenatedBinary(self, n: int) -> int:
        modulo = 10 ** 9 + 7
        shift = 0 # tracking power of 2
        res = 0
        
        for i in range(1, n+1):
            if i & (i - 1) == 0: # see if num reaches a greater power of 2
                shift += 1
            res = ((res << shift) + i) % modulo # do the simulation
        
        return res